packages/system/src/Stream/Stream/gen.ts

```diff
@@ -25,21 +27,24 @@ export function gen<Eff extends GenStream<any, any, any>, AEff>(
   }) => Generator<Eff, AEff, any>
 ): Stream<_R<Eff>, _E<Eff>, AEff> {
   return suspend(() => {
-    const iterator = f(adapter as any)
-    const state = iterator.next()
     function run(
-      state: IteratorYieldResult<Eff> | IteratorReturnResult<AEff>
+      replayStack: any[]
     ): Stream<any, any, AEff> {
+      const iterator = f(adapter as any)
+      let state = iterator.next()
+      for(let i = 0; i < replayStack.length; i++){
+        if(state.done){
+          return fromEffect(die("Something very wrong has happened. Replaying values resulted in a premature end of the generator execution. Provided generator should be pure and perform effects only by yielding them, so that the generator can safely be re-run without side effects."))
```

Comment:
Better to make a dedicated Tagged Exception maybe in the GlobalExceptions module